### PR TITLE
Getregion opts

### PR DESCRIPTION
**Pull request by Shougo**
Continues of https://github.com/vim/vim/pull/13998

* Add opts arguments

* `{pos1}` and `{pos2}` can be list

~~I don't understand what is the correct argument check.~~

Note: I want to add `winid` support, but I gave up.  current routine depends on current buffer...

### CODE REVIEW

File: src/evalfunc.c
Comment (zeertzjq) [inline src/evalfunc.c]: It seems that `var2fpos()` is more often used for functions that get some value (e.g. `col()`, `virtcol()`), while `list2fpos()` is used for functions that change cursor position (e.g. `cursor()`). Also, `curswant` isn't used in this function, so `var2fpos()` behavior is less confusing.

Comment (Shougo) [reply src/evalfunc.c]: Oh, I have not know `var2fpos()` support list...

**Comment by chrisbra** [comment]
Thanks, It's a bit disappointing, that we first make it using string args and now we add in addition list args, plus an optional dict type. I thought we could avoid having different types for the same function. Also, initially when I suggested the `type` arg, I meant it to be using a dict object instead. Now we are adding yet another dict argument for additional settings. 

Let's see what others think.

**Comment by Shougo** [comment]
I have decided the 3rd argument `{type}` is not optional.
Because it is not only for visual mode function.  It should be explicit.

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
		When {pos1} or {pos2} is |List|, with two or three items:
```

**Comment by yegappan** [comment]
> Thanks, It's a bit disappointing, that we first make it using string args and now we add in addition list args, plus an optional dict type. I thought we could avoid having different types for the same function. Also, initially when I suggested the `type` arg, I meant it to be using a dict object instead. Now we are adding yet another dict argument for additional settings.
> 
> Let's see what others think.

I agree.  In one of the comments for the previous PR, I mentioned this point about using a Dict
for the third argument.  The third argument should be a Dict.

**Comment by yegappan** [comment]
> I have decided the 3rd argument `{type}` is not optional. Because it is not only for visual mode function. It should be explicit.

You can use a default behavior.  You can add a "visualmode" item in the Dict that specifies the desired visual mode.

File: src/evalfunc.c
Comment (yegappan) [inline src/evalfunc.c]: The arg3_string type check will no longer work this.  You need to have a special function that checks whether
the first argument is a String or a List and the fourth optional argument is a Dict.  In Vim9, supporting different
types for the same argument is deprecated.  From a type checking perspective, it is better to support only
one type for the arguments.

File: src/testdir/test_visual.vim
Comment (yegappan) [inline src/testdir/test_visual.vim]: Can you add a test for using `offset` in the List argument?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by chrisbra** [comment]
> I agree. In one of the comments for the previous PR, I mentioned this point about using a Dict
for the third argument. The third argument should be a Dict.

We added `getregion()` in `9.1.120` just last week. I suppose it is okay to change the `getregion()` behaviour and just use 

```
getregion({pos1}, {pos2}, {Dict})			*getregion()*
		Returns the list of strings from {pos1} to {pos2} as if it's
		selected in visual mode of {type}. {pos1} and {pos2} are list
                items. For possible values see |getpos()|
		{Dict} is a dictionary with additional properties and supports
                the following keys:
                'type':  specify the selection type, supports the following
                values: "v" for |characterwise| mode
			"V" for |linewise| mode
			"<CTRL-V>" for |blockwise-visual| mode
                'selection': specify the 'selection' mode using the values
                             from the 'selection' setting.
```

Something like this?

**Comment by yegappan** [comment]
> > I agree. In one of the comments for the previous PR, I mentioned this point about using a Dict
> > for the third argument. The third argument should be a Dict.
> 
> We added `getregion()` in `9.1.120` just last week. I suppose it is okay to change the `getregion()` behaviour and just use
> 
> ```
> getregion({pos1}, {pos2}, {Dict})			*getregion()*
> 		Returns the list of strings from {pos1} to {pos2} as if it's
> 		selected in visual mode of {type}. {pos1} and {pos2} are list
>                 items. For possible values see |getpos()|
> 		{Dict} is a dictionary with additional properties and supports
>                 the following keys:
>                 'type':  specify the selection type, supports the following
>                 values: "v" for |characterwise| mode
> 			"V" for |linewise| mode
> 			"<CTRL-V>" for |blockwise-visual| mode
>                 'selection': specify the 'selection' mode using the values
>                              from the 'selection' setting.
> ```
> 
> Something like this?

Yes.  That looks good to me.

**Comment by Shougo** [comment]
OK.  I have changed `{type}` to `{opts}`.

File: src/evalfunc.c
Comment (Shougo) [reply src/evalfunc.c]: Fixed.

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
getregion({pos1}, {pos2} [, {opts}])
```

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
getregion({pos1}, {pos2} [, {opts}])			*getregion()*
```

File: src/evalfunc.c
Comment (zeertzjq) [inline src/evalfunc.c]: I don't think it's a good idea to use `VIsual_mode` as the default type, as it seems that changing the motion force of an operator can also change `VIsual_mode`. I think it's better to have a deterministic default value "v".

**Comment by Shougo** [comment]
I have changed `{type}` default value.

**Comment by Shane-XB-Qian** [comment]
> In Vim9, supporting different types for the same argument is deprecated.
> From a type checking perspective, it is better to support only
> one type for the arguments.

i may suggest re-thinking this "claim", IMO:
1. user defined func perhaps should be only one type for an argument,
2. but to vim native function "supporting different types for the same argument" was a tricky/classic/good design.

-- 
shane.xb.qian

**Comment by chrisbra** [comment]
Thanks, but I'd rather see either we change the first two position arguments to a list type and drop the string type or we do not change the type further. But let's not do both. It's also better for Vim9 type checking when compiling the function.

**Comment by Shougo** [comment]
> Thanks, but I'd rather see either we change the first two position arguments to a list type and drop the string type or we do not change the type further. But let's not do both. It's also better for Vim9 type checking when compiling the function.

I have changed the interface.  `{string}` is not allowed.

**Comment by chrisbra** [comment]
Can I get some comments before I merge? Everybody fine with this backwards-incompatible change?

**Comment by habamax** [comment]
> Can I get some comments before I merge? Everybody fine with this backwards-incompatible change?

I think it is fine, I doubt anyone used `getregion()` in the plugins yet.

File: runtime/doc/builtin.txt
Comment (habamax) [inline runtime/doc/builtin.txt]: In this example '.' is a string representing current position? But pos1 couldn't be a string, right?

File: runtime/doc/builtin.txt
Comment (habamax) [inline runtime/doc/builtin.txt:4283]: Should this line be removed?

File: runtime/doc/builtin.txt
Comment (habamax) [inline runtime/doc/builtin.txt]: `{type}` is in the dict now, and here it is referenced as a standalone parameter of a function.

File: runtime/doc/builtin.txt
Comment (Shougo) [reply runtime/doc/builtin.txt]: Fixed.

File: runtime/doc/builtin.txt
Comment (Shougo) [reply runtime/doc/builtin.txt:4283]: Yes.  I have removed.

**Comment by Shougo** [comment]
I have fixed the documentation.

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]: I don't think "off" is ignored. `coladd` is used in the code.

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]: There is no mention of "mark" anywhere else in the docs of this function, and it doesn't allow non-current buffer.

File: runtime/doc/builtin.txt
Comment (Shougo) [reply runtime/doc/builtin.txt]: Oh, I forgot it.

**Comment by zeertzjq** [comment]
Actually now that `pos1` and `pos2` only support lists, it should be possible to support non-current buffer as long as they have the same buffer number.

**Comment by Shougo** [comment]
> Actually now that pos1 and pos2 only support lists, it should be possible to support non-current buffer as long as they have the same buffer number.

It is possible.  But I gave up.  The PR is welcome.

File: runtime/doc/builtin.txt
Comment (zeertzjq) [inline runtime/doc/builtin.txt]: I think it's better to simply reference `getpos()`
```suggestion
		{pos1} and {pos2} must be a |List| with four numbers in the
		same format as the result of |getpos()|:
			[bufnum, lnum, col, off]
		While "bufnum" must be zero or current buffer number.
```

File: src/evalfunc.c
Comment (zeertzjq) [inline src/evalfunc.c]: `curswant` is not used and `list2fpos()` can accept `NULL` as `curswantp`
```suggestion
```

File: src/evalfunc.c
Comment (zeertzjq) [inline src/evalfunc.c]:
```suggestion
    if (list2fpos(&argvars[0], &p1, &fnum, NULL, FALSE) != OK
```

File: src/evalfunc.c
Comment (zeertzjq) [inline src/evalfunc.c]:
```suggestion
    if (list2fpos(&argvars[1], &p2, &fnum, NULL, FALSE) != OK
```

**Review by zeertzjq — COMMENTED** [review]
I think some of the tests should omit the `opts` argument or the `type` field to check for the default `type` behavior.

**Comment by Shougo** [comment]
Fixed.

**Review by Shane-XB-Qian — COMMENTED** [review]
- Order of {pos1} and {pos2} doesn't matter, it will always
		  return content from the upper left position to the lower
		  right position.

i doubt this design, tho it is useful, but similar vim native function e.g prop_remove was not playing like that, should keep consistent?

**Comment by Shougo** [comment]
Well, `prop_remove()` arguments are line numbers.  It seems `getline()` like interface.

Of course, I can change the behavior.  If pos1 pos2 is reversed, it can returns empty.

**Comment by Shougo** [comment]
I have checked visual mode region behavior.  It works if visual start and cursor position is swapped.

**Comment by Shane-XB-Qian** [comment]
yes, please think if it should 'order' matters, 
and actually if it was a visual selection, really needed to pass pos1/2 ?

**Comment by Shougo** [comment]
Because it is not only for visual mode.  It is just get region text from `pos1` to `pos2`.

**Comment by Shane-XB-Qian** [comment]
> `getregion()` <- what?

maybe `getregiion(vs_type)` was enough?

**Comment by habamax** [comment]
In case of getting visual text in vimscript I don't like the idea of checking if pos1 is less than pos2 before calling the function. I often use `o` to extend visual from both sides...

**Comment by chrisbra** [comment]
> yes, please think if it should 'order' matters,

Order should not matter. Let's not add burden to plugin authors so everybody has to figure out how to determine which position comes first. That is not so trivial

> and actually if it was a visual selection, really needed to pass pos1/2 ?

Yes, let's please keep it consistent. It should be fine to always give two position as arguments. The function signature should be the same always. That's better for Vim9 scripting it can easier detect potential problems.

**Comment by Shane-XB-Qian** [comment]
> > yes, please think if it should 'order' matters,
> 
> Order should not matter. Let's not add burden to plugin authors so everybody has to figure out how to determine which position comes first. That is not so trivial

yes, that's the why asking such "question", since now some native function was 'order' matter.

> > and actually if it was a visual selection, really needed to pass pos1/2 ?
> 
> Yes, let's please keep it consistent. It should be fine to always give two position as arguments. The function signature should be the same always. That's better for Vim9 scripting it can easier detect potential problems.

and further, actually to this 'getregion()' function, looked the pos1/2 not only not matter, but also seems maybe not required, or the signature should be `getregion(vs_type, [pos1, [pos2]])`, that maybe was the reducing plugin author or general vim user's burden.
// just my opinion, but your call.

**Comment by chrisbra** [comment]
> yes, that's the why asking such "question", since now some native function was 'order' matter.

For the purpose of getting the region, we only need to know the beginning and end-points. Order shouldn't matter, Vim can figure it out on its own which one is the starting and which one is the end position, right? For all other examples, this is not the case and then order matters of course.

> and further, actually to this 'getregion()' function, looked the pos1/2 not only not matter, but also seems maybe not required, or the signature should be getregion(vs_type, [pos1, [pos2]]), that maybe was the reducing plugin author or general vim user's burden.

I would think the positions are the most important parameters, so they should come first, don't you think? Then we have added another parameter, that allows you to further specify additional properties but you don't have to specify them, it's an optional dictionary. And we can later add additional settings, without changing the function signature.

**Comment by Shane-XB-Qian** [comment]
> I would think the positions are the most important parameters, so they should come first, don't you think? 

i donot know, but please check the signature of prop_remove and prop_clear, that maybe give your answer, 
to me, i felt getregion is similar like prop_remove.
// all in all, i think the usage of vim should keep consistent, specially those tricky/classic/good design.

**Comment by chrisbra** [comment]
> i donot know, but please check prop_remove and prop_clear, that maybe give your answer, to me, i felt getregion is similar like prop_remove.

I don't think so, it is quite different.

**Comment by yegappan** [comment]
> I would think the positions are the most important parameters, so they should come first, don't you think? Then we have added another parameter, that allows you to further specify additional properties but you don't have to specify them, it's an optional dictionary. And we can later add additional settings, without changing the function signature.
>

Yes.  If a builtin function needs optional arguments, then it should use a Dict as the last argument.
Instead of a Dict, if a specific type is used for the optional arguments, then when using multiple optional
arguments, to specify an optional argument all the preceding optional arguments must be specified.
For example, take the count() function as an example:

   count({comp}, {expr} [, {ic} [, {start}]])

This function accepts two optional arguments.  To specify the starting position, the 'ignorecase'
argument must be specified (even if it is not needed).  This is not a good design.

**Comment by Shane-XB-Qian** [comment]
> For example, consider the count() function:
> 
> count({comp}, {expr} [, {ic} [, {start}]])
> 
> This function accepts two optional arguments. To specify the starting position, the 'ignorecase' argument must be specified (even if it is not needed). This is not a good design. The last argument should have been a Dict with "ignorecase" and "startidx" items.

not same case, here pos1/2 was some kind of burden:
`to this 'getregion()' function, looked the pos1/2 not only 'order' not matter, but also seems maybe not required`

**Comment by chrisbra** [comment]
> to this 'getregion()' function, looked the pos1/2 not only 'order' not matter, but also seems maybe not required

it may only be not required, if you want to get the last visual selected region. But what if you want to get the region for various other marks?

**Comment by Shane-XB-Qian** [comment]
> > to this 'getregion()' function, looked the pos1/2 not only 'order' not matter, but also seems maybe not required
> 
> it may only be not required, if you want to get the last visual selected region. But what if you want to get the region for various other marks?

then what that 3rd arg `type` default `v` for?
or was there really possible needing for `other marks`? or had `getbufline` there for them.

-- 
shane.xb.qian

**Comment by chrisbra** [comment]
> then what that 3rd arg `type` default `v` for?

To specify what kind of a selection you want? That matters for all cases of selection of course. 

> or was there really possible needing for `other marks`? or had `getbufline` there for them.

Why not? Why are you trying to argue so much against it?

**Comment by Shane-XB-Qian** [comment]
> > then what that 3rd arg `type` default `v` for?
> 
> To specify what kind of a selection you want? That matters for all cases of selection of course. 

ok, so here 'selection' meant selection shape, not visual selection.

> > or was there really possible needing for `other marks`? or had `getbufline` there for them.
> 
> Why not? Why are you trying to argue so much against it?

i am not against it, or what benefit i would get if against it? :smile:

just the usage was confusing, and doubt the design or if "against" the passed vim classic design;
and to 'visual selection' the pos1/pos2 was redundant.

there may should be two func:
1. getregion(pos1, pos2, {opts})
2. getselection({opts})

or just similar one:
a-1. getregion({opts}, [pos1, [pos2]])
a-2. getregion({opts}, [pos1, pos2])

-- 
shane.xb.qian

**Comment by chrisbra** [comment]
> if "against" the passed vim classic design

So what exaclty was vim classic design? We have always added new features. 

> 1. getregion(pos1, pos2, {opts})
> 2. getselection({opts})

This adds too much clutter. Why having 2 functions performing almost the same thing?

> a-1. getregion({opts}, [pos1, [pos2]])
> a-2. getregion({opts}, [pos1, pos2])

It doesn't make sense to `{opts}` as the first variable, since it is optional.  We could have made one list of position lists, but I think then it wouldn't work as `-> method` call

**Comment by Shane-XB-Qian** [comment]
> > a-1. getregion({opts}, [pos1, [pos2]])
> > a-2. getregion({opts}, [pos1, pos2])
> 
> It doesn't make sense to `{opts}` as the first variable, since it is optional.  We could have made one list of position lists, but I think then it wouldn't work as `-> method` call

please check 'prop_remove()', similar situation, it can be `-> method` call, if `a-1` not preferred, then perhaps `a-2` was good.
and {opts} was not optional, since had to indicate a 'selection shape' anyhow as above said so, just there was a default value.

-- 
shane.xb.qian

**Comment by Shane-XB-Qian** [comment]
but it is truly Not worthy to refactor it again, 
it shoud be ok as long as if user knew that was not 'visual selection' but 'selection shape'.. :shrug:
let it be~ thanks arguing with me.. :smile:

**Pull request closed without merging**
